### PR TITLE
Use sedlex instead of ulex

### LINKED ISSUE #1792: Migrate from ulex to sedlex
Long story short - ulex is dying, mainstream depends on the camlp4, which is dead, fork depends on the camlp5, which seems dead too. This prevents from porting F* to more modern OCaml versions and dependencies versions.

See these links for more information:
- https://discuss.ocaml.org/t/camlp5-and-ocaml-4-08/3985/4
- https://github.com/ocaml/opam-repository/pull/14351

Sedlex is an active maintained project by OCaml community: https://github.com/ocaml-community/sedlex

### PR DESCRIPTION
**Pull request by mnxn**
Closes #1792

Based off of @fangyi-zhou's branch: https://github.com/fangyi-zhou/FStar/tree/sedlex

Moving to sedlex removes the transitive camlp4 dependency and allows F* to compile on recent OCaml versions (I tested on 4.11).

Most of the changes were straightforward syntax differences between ulex and sedlex, but there are a few noteworthy points:

- Sedlex doesn't provide an equivalent `Utf8` module, so I just copied in ulex's [`utf8.ml`](https://github.com/whitequark/ulex/blob/master/utf8.ml) as `FStar_Parser_Utf8.ml` (license embedded in file). It could probably be rewritten differently, but I wanted to avoid changes in logic outside of the lexer.

- Compilation of FStar_Parser_LexFStar.ml hangs if op_token is represented as one regexp. I believe this is due to its large size. Splitting op_token up into 5 regexp parts appears to fix the problem. Possibly related to ocaml-community/sedlex#34

- An underscore in sedlex works differently than ulex/ocamllex as it can match an empty string (see ocaml-community/sedlex#51). It is important to use the `any` keyword instead to match a single character. I have  done this in the `string` and `comment` lexer rules.

- Sedlex rules require a catch-all at the end of the rule. Rules that didn't have this case now include `| _ -> assert false`. I'm fairly certain that this case is impossible for the `string` and `comment` rules, but I am unsure about `one_line_comment` and `symbolchar_parser`. Is keeping the `assert false` fine or should it raise a different exception?

I ran `make -C tests/micro-benchmarks` and `make -C ulib -j6` successfully with the sedlex lexer. `make bench` has the following results:

**This PR - sedlex, OCaml 4.11:**

| | N_benches | total | user | system | mem(mb) |
| - | - | - | - | -| - |
| micro-benchmarks | 229 | 57.6017 | 37.341 | 15.7662 | 23521214 |
| ulib | 288 | 1833.95 | 1764.36 | 48.22 | 73497874 |
| ocaml_extract | 95 | 140.748 | 125.184 | 13.6787 | 18332827 |

**Master - ulex, OCaml 4.09:**

| | N_benches | total | user | system | mem(mb) |
| - | - | - | - | -| - |
| micro-benchmarks | 229 | 60.1398 | 39.9449 | 15.6857 | 23607410 |
| ulib | 288 | 1799.14 | 1714.82 | 43.023 | 74029871 |
| ocaml_extract | 95 | 135.334 | 121.343 | 12.6106 | 18295119 |

If I'm understanding these results correctly, the F* compiler is slightly slower with sedlex but uses less memory. The difference is small so it probably has more to do with my computer and the different OCaml compiler version.

Submission containing materials of a third party:

- @fangyi-zhou, base of sedlex rewrite
- Alain Frisch, UTF8 module, https://github.com/whitequark/ulex/blob/master/LICENSE

### CODE REVIEW

**Comment by nikswamy** [comment]
This is awesome! Thank you! Sorry for not noticing it earlier, but will try to merge it soon.

File: src/parser/ml/FStar_Parser_LexFStar.ml [hunk 1]
```diff
@@ -304,141 +304,146 @@ let push_one_line_comment pre lexbuf =
...
-  "~" | "-" | "/\\" | "\\/" | "<:" | "<@" | "(|" | "|)" | "#" |
-  "u#" | "&" | "()" | "(" | ")" | "," | "~>" | "->" | "<--" |
-  "<-" | "<==>" | "==>" | "." | "?." | "?" | ".[|" | ".[" | ".(|" | ".(" |
-  "$" | "{:pattern" | ":" | "::" | ":=" | ";;" | ";" | "=" | "%[" |
-  "!{" | "[@@" | "[@" | "[|" | "{|" | "[" | "|>" | "]" | "|]" | "|}" | "{" | "|" | "}"
+let op_char = [%sedlex.regexp? Chars "!$%&*+-./<=?^|~:"]
+let ignored_op_char = [%sedlex.regexp? Chars ".$"]
+
+(* op_token must be splt into seperate regular expressions to prevent
+   compliation from hanging *)
```
Comment (nikswamy) [inline src/parser/ml/FStar_Parser_LexFStar.ml:384]: that's interesting ... can you comment about this?

File: src/parser/ml/FStar_Parser_LexFStar.ml [hunk 2]
```diff
@@ -502,30 +506,35 @@ let rec token = lexer
...
- | '`' '`' (([^'`' 10 13 0x2028 0x2029] | '`' [^'`' 10 13 0x2028 0x2029])+) '`' '`' ->
+ | '`', '`', (Plus (Compl ('`' | 10 | 13 | 0x2028 | 0x2029) | '`', Compl ('`' | 10 | 13 | 0x2028 | 0x2029))), '`', '`' ->
    IDENT (trim_both lexbuf 2 2)
 
- | op_token  -> L.lexeme lexbuf |> Hashtbl.find operators
+ | op_token_1
+ | op_token_2
+ | op_token_3
+ | op_token_4
+ | op_token_5 -> L.lexeme lexbuf |> Hashtbl.find operators
```
Comment (nikswamy) [inline src/parser/ml/FStar_Parser_LexFStar.ml:518]: Using a single large regexp is less efficient than the disjunction of five smaller ones?

File: src/parser/ml/FStar_Parser_LexFStar.ml (hunk 1 above)
Comment (nikswamy) [reply src/parser/ml/FStar_Parser_LexFStar.ml:384]: How did you diagnose that this particular regexp was the source of the hanging problem?

Comment (nikswamy) [reply src/parser/ml/FStar_Parser_LexFStar.ml:384]: I indeed see it hanging on

00:00:58 88   (87  ) ...parser/ml/FStar_Parser_LexFStar.ml.depends   O-b----- \
00:00:58 88   (87  ) ...parser/ml/FStar_Parser_LexFStar.ml.depends   O-b----- \
00:00:59 88   (87  ) ...parser/ml/FStar_Parser_LexFStar.ml.depends

Comment (mnxn) [reply src/parser/ml/FStar_Parser_LexFStar.ml:384]: > How did you diagnose that this particular regexp was the source of the hanging problem?

It was mostly just trial and error changing things until the code compiled :)

I am not familiar with sedlex internals so I don't know the exact reason for this problem. When I have some free time, I will try to create a minimal repro and ask the sedlex maintainers.

Comment (mnxn) [reply src/parser/ml/FStar_Parser_LexFStar.ml:384]: I opened an issue in the sedlex repository here: https://github.com/ocaml-community/sedlex/issues/97

Comment (nikswamy) [reply src/parser/ml/FStar_Parser_LexFStar.ml:384]: thank you! This is *very* helpful

**Comment by catalin-hritcu** [comment]
@mnxn You mention that
> Moving to sedlex removes the transitive camlp4 dependency and allows F* to compile on recent OCaml versions (I tested on 4.11).

I agree with the first part, but I'm not sure at all about the second part. I tried to build F* now on OCaml 4.10, 4.11, and 4.12, and they all failed for me because of the the ocaml-migrate-parsetree < 2.1.0 constraint. More details in #2246.

**Pull request merged**